bson-codec-design/DESIGN.md

```diff
@@ -0,0 +1,157 @@
+# BSON Encoder & Decoder Design
+The v1 design for the BSON library's encoder and decoder contain design issues that
+require a redesign. The main design flaw is the lack of a coherent, flexible design that can
+incorporate the needs of users. Assumptions made during the v1 design have proven incorrect or too
+inflexible to be as useful as initially intended.
+
+## Definitions
+<dl>
+<dt>Codec Processing</dt>
+<dd>The complete process of either encoding or decoding.</dd>
+<dt>Unmarshal</dt>
+<dd>The process of converting a slice of bytes into a native Go type</dd>
+<dt>Marshal</dt>
+<dd>The process of converting a native Go type into a slice of bytes</dd>
+<dt>Encode</dt>
+<dd>The process of converting a native Go type into BSON and writing it to an
+<code>io.Writer</code></dd>
+<dt>Decode</dt>
+<dd>The process of reading a BSON document from an <code>io.Reader</code> and converting it to a
+native Go type</dd>
+<dt>V1 Design </dt>
+<dd>The initial design for the encoding and decoding functionality of the BSON library</dd>
+<dt>Proposed Design</dt>
+<dd>The design laid out in this document</dd>
+</dl>
+
+## Motivation
+The v1 design for codec processing in the BSON library made a number of assumptions about the
+control the user had over their types, the amount of granular control users would want to have over
+codec processing, and the flexibility required during codec processing. A key initial assumption of
+the v1 design was that users who needed to handle types in a custom way would be able to make those
+types implement the `bson.Marshaler` and `bson.Unmarshaler` interfaces. This turned out to be too
+restrictive because users sometimes want to handle the processing for types they do not define
+themselves. For example, users who generate types using a protocol buffer library
+or users who want to handle the processing of types in the standard library.
+
+The design of the v1 encoder and decoder for the Go driver presents problems when users wish to
+have more control over how specific types are handled when encoding or decoding. The control that
+users require exists on two main levels:
+- control over how individual types are encoded or decoded
+- control over how tags are processed
+
+The granularity of control users require can be defined in three levels:
+- Control over how individual types are encoded and decoded when they are embedded in another type,
+such as a map or a struct
+- Control over how individual types are encoded and decoded when they are provided directly to the
+encoder or decoder
+- Control over how tags are processed and the parameters that are returned
+
+The proposed design allows for the first two granularities to be handled together.
+
+## High Level Design
```

Comment:
I've left a few specific comments but: 

- will there be a default registry that users will be able to create?
- it's not obvious to me why interfaces and types need different registries. 
- we should address recursiveness explicitly in this document, because it's confusing, and a higher level bug in v1. How do we use. 
- while Decoders are logically stateless, we're going to be reusing decoders for multiple operations. as a result we should add some notes about expectations for threadsafety, or their ability to use sync.Pools to manage decoders/registries. 
- if types implement Marshaller/Unmarshaller and are *also* registered, should we use the rules from the registry or let the marshaller win? should this be an error?

Reply:
> > we should address recursiveness explicitly in this document, because it's confusing, and a higher level bug in v1. How do we use.
>
> Can you clarify this question?

The v1 decoder only looks if the top level type implements unmarsheler, if a struct does not implement unmarshaller (and maybe even if it does) but one of its fields *does* implement unmarshaller, those unmarshal functions are never called. The v2 decoder should support unmarshallers at all levels. 

> So, to answer this question, whether marshaller/unmarshaller is in the interface registry, we should pretend it is and abide by that logic.

I'm totally fine with your logic here, but think that we should be explicit about this and call this semantic out in this document as well as the final documentation.

Reply:
> The v1 decoder only looks if the top level type implements unmarsheler, if a struct does not implement unmarshaller (and maybe even if it does) but one of its fields does implement unmarshaller, those unmarshal functions are never called. The v2 decoder should support unmarshallers at all levels.

That's a bug, one that we haven't fixed because we're implementing a new design. Also, I'm not sure how one would use the registry for everything and somehow to forget to make it recursive. I suppose it could be a footnote of some sort.

Reply:
> I'm totally fine with your logic here, but think that we should be explicit about this and call this semantic out in this document as well as the final documentation.

> For instance, the bson package contains the bson.Marshaler and bson.Unmarshaler interfaces. Since an interface is less specific than a type, the Interface Registry is consulted after the Type registry. Therefore if type implements and interface and there is a codec registered for the type in the Type Registry and there is a codec registered for the interface in the Interface Registry, the codec in the Type Registry will be used.

Previously this was stated as:
> For instance, the bson package contains the bson.Marshaler and bson.Unmarshaler interfaces. When a type implements one of these interfaces, the codec processing is delegated to a codec registered in the Interface Registry, not the Type Registry.

After an offline discussion with @craiggwilson we swapped the ordering around, but this ordering has been here since the initial draft.

Reply:
No open issues from this thread.